File: application/app/gateway/backend/ConnectorResolver.scala
package gateway.backend

import domain.exceptions.BackendNotSupportedException
import domain.model.{Backend, BackendSettings, Route}
import javax.inject.{Inject, Singleton}

@Singleton
class ConnectorResolver @Inject() (private val connectors: Set[BackendConnector]) {

  def getConnectors(route: Route): Set[BackendConnector] = route.backends.map(getConnector)

  def getConnector(backend: Backend[BackendSettings]): BackendConnector = {
    connectors.find(c => c.supports(backend.backendType)) match {
      case Some(c) => c
      case None => throw BackendNotSupportedException(s"Backend of type ${backend.backendType} is not supported")
    }
  }
}

File: ws-gateway-application/src/main/java/com/cosmin/wsgateway/application/gateway/connector/ConnectorResolver.java
package com.cosmin.wsgateway.application.gateway.connector;

import com.cosmin.wsgateway.domain.Backend;
import com.cosmin.wsgateway.domain.BackendSettings;
import com.cosmin.wsgateway.domain.exceptions.BackendNotSupportedException;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;


@Component
@RequiredArgsConstructor
public class ConnectorResolver {
    private final List<BackendConnector<? extends BackendSettings>> connectors;

    public BackendConnector<? extends BackendSettings> getConnector(Backend<? extends BackendSettings> backend) {
        return connectors.stream()
                .filter(c -> c.supports(backend.type()))
                .findFirst()
                .orElseThrow(() -> new BackendNotSupportedException(backend.type().toString()));
    }
}
